### PR TITLE
Osgi fixes

### PR DESCRIPTION
**Pull request by arichiardi**
Hello Laurent,
I added the two fixes discussed in #769 and #767 (actually the ```synchronized``` comment is somewhere I cannot find anymore).

The first just removes ```synchronized``` from ```ClojureOSGi.initialize()```, the second solves the exception:

    !ENTRY org.eclipse.osgi 2 0 2015-05-17 10:08:59.544
    !MESSAGE State change in progress for bundle "reference:file:/home/***/git/ccw/ccw.core/" by thread "main".
    !STACK 0
    org.osgi.framework.BundleException: State change in progress for bundle "reference:file:/home/***/git/ccw/ccw.core/" by thread "main".
    at org.eclipse.osgi.framework.internal.core.AbstractBundle.beginStateChange(AbstractBundle.java:1088)
	at org.eclipse.osgi.framework.internal.core.AbstractBundle.start(AbstractBundle.java:298)
	at org.eclipse.osgi.framework.util.SecureAction.start(SecureAction.java:478)

I hope I have not messed up with it too much :wink:

### CODE REVIEW

File: ccw.core/src/java/ccw/CCWPlugin.java [hunk 1]
```diff
@@ -234,6 +234,17 @@ private final void logDependenciesInformation(final BundleContext context) {
...
 
+    private volatile int bundleState = Bundle.UNINSTALLED;
+
+    /**
+     * Return if the bundle is going to stop.
+     * @return True or false.
+     */
+    public static boolean isBundleStopping() {
+        CCWPlugin plugin = CCWPlugin.getDefault();
+        return plugin != null ? plugin.bundleState == Bundle.STOPPING : false;
```
Comment (laurentpetit) [inline ccw.core/src/java/ccw/CCWPlugin.java]: Shouldn't we refuse to load code with withBundle() is the plugin is null? Right now your change returns false if the plugin is null, and so withBundle() will continue its work.
Also, if the bundle is stopped, or uninstall, or I don't know what, maybe we shouldn't load code as well? Not just "while" it is stopping? => What about being more explicit in the test condition: if (isBundleStopping() || plugin == null) return null; ?

Comment (arichiardi) [inline ccw.core/src/java/ccw/CCWPlugin.java]: Yes probably I should change this method to some ```canLoadWithBundle``` because you are right, there are many more cases where we don't want to load...don't know why I didn't think about it ...

File: ccw.core/src/java/ccw/util/osgi/ClojureOSGi.java [hunk 2]
```diff
@@ -82,6 +81,10 @@ private static DynamicClassLoader getDynamicClassLoader(Bundle bundle, List<URL>
 	public static Object withBundle(Bundle aBundle, RunnableWithException aCode, List<URL> additionalURLs)
 			throws RuntimeException {
 
+		if (Thread.currentThread().isInterrupted() || CCWPlugin.canLoadCodeInBundle()) {
```
Comment (laurentpetit) [inline ccw.core/src/java/ccw/util/osgi/ClojureOSGi.java]: shouldn't the test be `!CCWPlugin.canLoadCodeInBundle()` ?

**Comment by laurentpetit** [comment]
I have rebased and changed the test in `withBundle()`. If everything goes well I will merge into master and then close this PR

File: ccw.core/src/java/ccw/util/osgi/ClojureOSGi.java (hunk 2 above)
Comment (arichiardi) [inline ccw.core/src/java/ccw/util/osgi/ClojureOSGi.java]: Yep sorry I am going to change!

Comment (laurentpetit) [inline ccw.core/src/java/ccw/util/osgi/ClojureOSGi.java]: no problem, already changed and pushed. Just asked to double check

2015-06-12 13:56 GMT+02:00 Andrea Richiardi <notifications@github.com>:

> In ccw.core/src/java/ccw/util/osgi/ClojureOSGi.java
> <https://github.com/laurentpetit/ccw/pull/782#discussion_r32309462>:
>
> > @@ -82,6 +81,10 @@ private static DynamicClassLoader getDynamicClassLoader(Bundle bundle, List<URL>
> >  	public static Object withBundle(Bundle aBundle, RunnableWithException aCode, List<URL> additionalURLs)
> >  			throws RuntimeException {
> >
> > +		if (Thread.currentThread().isInterrupted() || CCWPlugin.canLoadCodeInBundle()) {
>
> Yep sorry I am going to change!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/laurentpetit/ccw/pull/782/files#r32309462>.
>



-- 
Laurent Petit

**Comment by arichiardi** [comment]
Ok perfect!
